Feature modules integration

Hi, 

I'm migrating a pet project from a single module to multi modules (and instant app in the future) and I need JaCoCo integration. So, I tryied the snapshot version (0.12.0-SNAPSHOT) in the project but the following exception was thrown when running any command of Gradle:

```
Caused by: groovy.lang.MissingMethodException: No signature of method: java.util.ArrayList.all() is applicable for argument types: (com.vanniktech.android.junit.jacoco.GenerationPlugin$_addJacocoAndroid_closure7) values: [com.vanniktech.android.junit.jacoco.GenerationPlugin$_addJacocoAndroid_closure7@697ddda9]
Possible solutions: any(), tail(), tail(), last(), last(), add(java.lang.Object)
        at com.vanniktech.android.junit.jacoco.GenerationPlugin.addJacocoAndroid(GenerationPlugin.groovy:127)
        at com.vanniktech.android.junit.jacoco.GenerationPlugin.addJacoco(GenerationPlugin.groovy:48)
        at com.vanniktech.android.junit.jacoco.GenerationPlugin.addJacoco(GenerationPlugin.groovy:41)
        at com.vanniktech.android.junit.jacoco.GenerationPlugin$_apply_closure2.doCall(GenerationPlugin.groovy:35)
        at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:40)
        at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:25)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:230)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:149)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:324)
        at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:234)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:140)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
        at com.sun.proxy.$Proxy24.afterEvaluate(Unknown Source)
        at org.gradle.configuration.project.LifecycleProjectEvaluator.notifyAfterEvaluate(LifecycleProjectEvaluator.java:76)
```

Thanks

--- a/src/main/groovy/com/vanniktech/android/junit/jacoco/GenerationPlugin.groovy
+++ b/src/main/groovy/com/vanniktech/android/junit/jacoco/GenerationPlugin.groovy
@@ -43,8 +43,7 @@ class GenerationPlugin implements Plugin<Project> {
 
     protected static boolean addJacoco(final Project subProject, final JunitJacocoExtension extension, JacocoMerge mergeTask, JacocoReport mergedReportTask) {
         if (!shouldIgnore(subProject, extension)) {
-            if (isAndroidApplication(subProject) || isAndroidLibrary(subProject) ||
-                    isAndroidFeature(subProject) || isAndroidInstantApp(subProject)) {
+            if (isAndroidProject(subProject)) {
                 addJacocoAndroid(subProject, extension, mergeTask, mergedReportTask)
                 return true
             } else if (isJavaProject(subProject)) {
@@ -120,7 +119,7 @@ class GenerationPlugin implements Plugin<Project> {
         Collection<BaseVariant> variants = []
         if (isAndroidApplication(subProject)) {
             variants = subProject.android.applicationVariants
-        } else if (isAndroidLibrary(subProject)) {
+        } else {
             variants = subProject.android.libraryVariants
         }
 
@@ -293,28 +292,28 @@ class GenerationPlugin implements Plugin<Project> {
         ] : extension.excludes
     }
 
-    protected static boolean isAndroidLibrary(final Project project) {
-        return project.plugins.hasPlugin('com.android.library')
+    private static boolean isAndroidProject(final Project project) {
+        final boolean isAndroidLibrary = project.plugins.hasPlugin('com.android.library')
+        final boolean isAndroidApp = project.plugins.hasPlugin('com.android.application')
+        final boolean isAndroidTest = project.plugins.hasPlugin('com.android.test')
+        final boolean isAndroidFeature = project.plugins.hasPlugin('com.android.feature')
+        final boolean isAndroidInstantApp = project.plugins.hasPlugin('com.android.instantapp')
+        return isAndroidLibrary || isAndroidApp || isAndroidTest || isAndroidFeature || isAndroidInstantApp
     }
 
-    protected static boolean isAndroidApplication(final Project project) {
-        return project.plugins.hasPlugin('com.android.application')
-    }
-
-    protected static boolean isJavaProject(final Project project) {
-        return project.plugins.hasPlugin('org.gradle.java')
-    }
-
-    protected static boolean isAndroidInstantApp(final Project project) {
-        return project.plugins.hasPlugin('com.android.instantapp')
-    }
-
-    protected static boolean isAndroidFeature(final Project project) {
-        return project.plugins.hasPlugin('com.android.feature')
+    private static boolean isJavaProject(final Project project) {
+        final boolean isJava = project.plugins.hasPlugin('java')
+        final boolean isJavaLibrary = project.plugins.hasPlugin('java-library')
+        final boolean isJavaGradlePlugin = project.plugins.hasPlugin('java-gradle-plugin')
+        return isJava || isJavaLibrary || isJavaGradlePlugin
     }
 
     protected static boolean isKotlinAndroid(final Project project) {
         return project.plugins.hasPlugin('org.jetbrains.kotlin.android')
+    }
+
+    protected static boolean isAndroidApplication(final Project project) {
+        return project.plugins.hasPlugin('com.android.application')
     }
 
     private static boolean shouldIgnore(final Project project, final JunitJacocoExtension extension) {

--- a/src/test/groovy/com/vanniktech/android/junit/jacoco/GenerationPluginSpec.groovy
+++ b/src/test/groovy/com/vanniktech/android/junit/jacoco/GenerationPluginSpec.groovy
@@ -5,7 +5,7 @@ import spock.lang.Specification
 import spock.lang.Unroll
 
 final class GenerationPluginSpec extends Specification {
-  final static ANDROID_PLUGINS = ["com.android.application", "com.android.library", "com.android.test"]
+  final static ANDROID_PLUGINS = ["com.android.application", "com.android.library", "com.android.test", "com.android.feature"]
   final static COMPILE_SDK_VERSION = 27
   final static BUILD_TOOLS_VERSION = "27.0.1"
   final static APPLICATION_ID = "com.example"

--- a/src/test/groovy/com/vanniktech/android/junit/jacoco/GenerationTest.groovy
+++ b/src/test/groovy/com/vanniktech/android/junit/jacoco/GenerationTest.groovy
@@ -8,11 +8,10 @@ import org.junit.Test
 
 import static com.vanniktech.android.junit.jacoco.ProjectHelper.ProjectType.*
 
-public class GenerationTest {
-  def LANGUAGES = ["clojure", "groovy", "java", "kotlin", "scala"]
+class GenerationTest {
+    def LANGUAGES = ["clojure", "groovy", "java", "kotlin", "scala"]
 
-  @Test
-    public void addJacocoAndroidAppWithFlavors() {
+    @Test void addJacocoAndroidAppWithFlavors() {
         def androidAppProject = ProjectHelper.prepare(ANDROID_APPLICATION).withRedBlueFlavors().get()
 
         GenerationPlugin.addJacoco(androidAppProject, new JunitJacocoExtension())
@@ -20,8 +19,7 @@ public class GenerationTest {
         assertJacocoAndroidWithFlavors(androidAppProject)
     }
 
-    @Test
-    public void addJacocoAndroidLibraryWithFlavors() {
+    @Test void addJacocoAndroidLibraryWithFlavors() {
         def androidLibraryProject = ProjectHelper.prepare(ANDROID_LIBRARY).withRedBlueFlavors().get()
 
         GenerationPlugin.addJacoco(androidLibraryProject, new JunitJacocoExtension())
@@ -29,8 +27,7 @@ public class GenerationTest {
         assertJacocoAndroidWithFlavors(androidLibraryProject)
     }
 
-    @Test
-    public void addJacocoAndroidApp() {
+    @Test void addJacocoAndroidApp() {
         def androidAppProject = ProjectHelper.prepare(ANDROID_APPLICATION).get()
 
         GenerationPlugin.addJacoco(androidAppProject, new JunitJacocoExtension())
@@ -38,8 +35,7 @@ public class GenerationTest {
         assertJacocoAndroidWithoutFlavors(androidAppProject)
     }
 
-    @Test
-    public void addJacocoAndroidLibrary() {
+    @Test void addJacocoAndroidLibrary() {
         def androidLibraryProject = ProjectHelper.prepare(ANDROID_LIBRARY).get()
 
         GenerationPlugin.addJacoco(androidLibraryProject, new JunitJacocoExtension())
@@ -47,8 +43,7 @@ public class GenerationTest {
         assertJacocoAndroidWithoutFlavors(androidLibraryProject)
     }
 
-    @Test
-    public void addJacocoJava() {
+    @Test void addJacocoJava() {
         def javaProject = ProjectHelper.prepare(JAVA).get()
 
         GenerationPlugin.addJacoco(javaProject, new JunitJacocoExtension())
@@ -56,8 +51,7 @@ public class GenerationTest {
         assertJacocoJava(javaProject)
     }
 
-    @Test
-    public void jacocoVersion() {
+    @Test void jacocoVersion() {
         final def extension = new JunitJacocoExtension()
         extension.jacocoVersion = '0.7.6.201602180812'
         def androidAppProject = ProjectHelper.prepare(ANDROID_APPLICATION).get()
@@ -73,8 +67,7 @@ public class GenerationTest {
         assert javaProject.jacoco.toolVersion == extension.jacocoVersion
     }
 
-    @Test
-    public void ignoreProjects() {
+    @Test void ignoreProjects() {
         final def extension = new JunitJacocoExtension()
         final def projects = [
                 ProjectHelper.prepare(ANDROID_APPLICATION).get(),
@@ -89,8 +82,7 @@ public class GenerationTest {
         }
     }
 
-    @Test
-    public void ignoreProjectsPath() {
+    @Test void ignoreProjectsPath() {
         final def extension = new JunitJacocoExtension()
         final def projects = [
                 ProjectHelper.prepare(ANDROID_APPLICATION).get(),
@@ -105,8 +97,7 @@ public class GenerationTest {
         }
     }
 
-    @Test
-    public void ignoreProjectsRegexPath() {
+    @Test void ignoreProjectsRegexPath() {
         final def extension = new JunitJacocoExtension()
         final def projects = [
                 ProjectHelper.prepare(ANDROID_APPLICATION).get(),
@@ -121,8 +112,7 @@ public class GenerationTest {
         }
     }
 
-    @Test
-    public void ignoreProjectsRegexName() {
+    @Test void ignoreProjectsRegexName() {
         final def extension = new JunitJacocoExtension()
         final def projects = [
                 ProjectHelper.prepare(ANDROID_APPLICATION).get(),
@@ -136,8 +126,7 @@ public class GenerationTest {
         }
     }
 
-    @Test
-    public void ignoreProjectsWrongRegexName() {
+    @Test void ignoreProjectsWrongRegexName() {
         final def extension = new JunitJacocoExtension()
         final def projects = [
                 ProjectHelper.prepare(ANDROID_APPLICATION).get(),
@@ -151,8 +140,7 @@ public class GenerationTest {
         }
     }
 
-    @Test
-    public void androidAppBuildExecutesJacocoTask() {
+    @Test void androidAppBuildExecutesJacocoTask() {
         def androidAppProject = ProjectHelper.prepare(ANDROID_APPLICATION).get()
 
         GenerationPlugin.addJacoco(androidAppProject, new JunitJacocoExtension())
@@ -161,8 +149,7 @@ public class GenerationTest {
         assert taskDependsOn(androidAppProject.check, 'jacocoTestReportRelease')
     }
 
-    @Test
-    public void androidLibraryBuildExecutesJacocoTask() {
+    @Test void androidLibraryBuildExecutesJacocoTask() {
         def androidLibraryProject = ProjectHelper.prepare(ANDROID_LIBRARY).get()
 
         GenerationPlugin.addJacoco(androidLibraryProject, new JunitJacocoExtension())
@@ -171,8 +158,7 @@ public class GenerationTest {
         assert taskDependsOn(androidLibraryProject.check, 'jacocoTestReportRelease')
     }
 
-    @Test
-    public void javaBuildExecutesJacocoTask() {
+    @Test void javaBuildExecutesJacocoTask() {
         def javaProject = ProjectHelper.prepare(JAVA).get()
 
         GenerationPlugin.addJacoco(javaProject, new JunitJacocoExtension())
@@ -363,8 +349,7 @@ public class GenerationTest {
         return false
     }
 
-    @Test
-    public void getExcludesDefault() {
+    @Test void getExcludesDefault() {
         final def excludes = GenerationPlugin.getExcludes(new JunitJacocoExtension())
 
         assert excludes.size == 20
@@ -390,8 +375,7 @@ public class GenerationTest {
         assert excludes.contains('**/*AutoValue_*.*')
     }
 
-    @Test
-    public void getExcludesCustom() {
+    @Test void getExcludesCustom() {
         final def extension = new JunitJacocoExtension()
         extension.excludes = new ArrayList<>()
         extension.excludes.add("**/*.java")

--- a/src/test/groovy/com/vanniktech/android/junit/jacoco/JunitJacocoExtensionTest.groovy
+++ b/src/test/groovy/com/vanniktech/android/junit/jacoco/JunitJacocoExtensionTest.groovy
@@ -2,14 +2,13 @@ package com.vanniktech.android.junit.jacoco
 
 import org.junit.Test
 
-public class JunitJacocoExtensionTest {
-    @Test
-    public void defaults() {
-        def extension = new JunitJacocoExtension()
+class JunitJacocoExtensionTest {
+  @Test void defaults() {
+    def extension = new JunitJacocoExtension()
 
-        assert extension.jacocoVersion == '0.7.2.201409121644'
-        assert extension.ignoreProjects.size() == 0
-        assert extension.excludes == null
-        assert !extension.includeNoLocationClasses
-    }
+    assert extension.jacocoVersion == '0.7.2.201409121644'
+    assert extension.ignoreProjects.size() == 0
+    assert extension.excludes == null
+    assert !extension.includeNoLocationClasses
+  }
 }

--- a/src/test/groovy/com/vanniktech/android/junit/jacoco/ProjectHelper.groovy
+++ b/src/test/groovy/com/vanniktech/android/junit/jacoco/ProjectHelper.groovy
@@ -10,15 +10,15 @@ import org.gradle.testfixtures.ProjectBuilder
 
 /** Provides projects for testing */
 final class ProjectHelper {
-    public static ProjectHelper prepare(ProjectType projectType) {
+    static ProjectHelper prepare(ProjectType projectType) {
         return new ProjectHelper(projectType)
     }
 
-    private final ProjectType projectType;
-    private final Project project;
+    private final ProjectType projectType
+    private final Project project
 
     private ProjectHelper(ProjectType projectType) {
-        this.projectType = projectType;
+        this.projectType = projectType
 
         switch (projectType) {
             case ProjectType.JAVA:
@@ -72,7 +72,7 @@ final class ProjectHelper {
     }
 
     /** Adds flavors to project, only for Android based projects */
-    public ProjectHelper withRedBlueFlavors() {
+    ProjectHelper withRedBlueFlavors() {
         if (projectType == ProjectType.JAVA) {
             throw new UnsupportedOperationException('Not supported with Java project')
         }
@@ -112,19 +112,19 @@ final class ProjectHelper {
         return this
     }
 
-    public Project get() {
+    Project get() {
         return project
     }
 
-    public enum ProjectType {
+    enum ProjectType {
         ANDROID_APPLICATION('com.android.application'),
         ANDROID_LIBRARY('com.android.library'),
-        JAVA('java');
+        JAVA('java')
 
-        private final String pluginName;
+        private final String pluginName
 
         ProjectType(String pluginName) {
-            this.pluginName = pluginName;
+            this.pluginName = pluginName
         }
     }
 }

